MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -212,208 +217,343 @@ double GeoMapper::getMeshElevation(
+
+    return sub_segments;
 }
 
-unsigned GeoMapper::getPointPosInLine(GeoLib::Polyline const*const line, unsigned start, unsigned end, GeoLib::Point const*const point, double eps) const
+static void mapPointOnSurfaceElement(MeshLib::Element const& elem,
+                                     MathLib::Point3d& q)
 {
-    const double* first = line->getPoint(start)->getCoords();
-    const double* pnt   = point->getCoords();
-    const double max_dist = MathLib::sqrDist(first, pnt);
+    // create plane equation: n*p = d
+    MathLib::Vector3 const& p(*(elem.getNode(0)));
+    MathLib::Vector3 const n(MeshLib::FaceRule::getSurfaceNormal(&elem));
+    if (n[2] == 0.0) { // vertical plane, z coordinate is arbitrary
+        q[2] = p[2];
+    } else {
+        double const d(MathLib::scalarProduct(n, p));
+        q[2] = (d - n[0]*q[0] - n[1]*q[1])/n[2];
+    }
+}
 
-    // if point is at start or end of line segment
-    if (max_dist<eps && MathLib::sqrDist(pnt, line->getPoint(end)->getCoords())) return 0;
+static std::vector<MeshLib::Element const*>
+getCandidateElementsForLineSegmentIntersection(
+    MeshLib::MeshElementGrid const& mesh_element_grid,
+    GeoLib::LineSegment segment)
+{
+    // modify z coordinates such that all surface elements around the line
+    // segment are found
+    segment.getBeginPoint()[2] = mesh_element_grid.getMinPoint()[2];
+    segment.getEndPoint()[2] = mesh_element_grid.getMaxPoint()[2];
+    std::array<MathLib::Point3d, 2> const pnts{
+        {segment.getBeginPoint(), segment.getEndPoint()}};
+    GeoLib::AABB aabb(pnts.cbegin(), pnts.cend());
 
-    for (std::size_t i=start+1; i<end; ++i)
-    {
-        const double* current = (*line->getPoint(i)).getCoords();
-        if (MathLib::sqrDist(pnt, current) < eps) return 0;
-        if (MathLib::sqrDist(first, current) > max_dist) return i;
+    auto candidate_elements = mesh_element_grid.getElementsInVolume(
+        aabb.getMinPoint(), aabb.getMaxPoint());
+
+    // make candidate elements unique
+    BaseLib::makeVectorUnique(candidate_elements);
+
+    return candidate_elements;
+}
+
+static bool snapPointToElementNode(MathLib::Point3d& p,
+                                   MeshLib::Element const& elem, double rel_eps)
+{
+    // values will be initialized within computeSqrNodeDistanceRange
+    double sqr_min, sqr_max;
+    elem.computeSqrNodeDistanceRange(sqr_min, sqr_max);
+
+    double const sqr_eps(rel_eps*rel_eps * sqr_min);
+    for (std::size_t k(0); k<elem.getNumberOfNodes(); ++k) {
+        auto const& node(*elem.getNode(k));
+        double const sqr_dist_2d(MathLib::sqrDist2d(p, node));
+        if (sqr_dist_2d < sqr_eps) {
+#ifndef NDEBUG
+            std::stringstream out;
+            out.precision(std::numeric_limits<double>::digits10);
+            out << "Segment point snapped from " << p;
+#endif
+            p = node;
+#ifndef NDEBUG
+            out << "to " << p;
+            DBUG("%s", out.str().c_str());
```

Comment:
Not sure here, but if this output is being showed in debug build of the testrunner, then better to comment it...